### PR TITLE
[IMP] account_python_tax: use ast to validate tax

### PR DESCRIPTION
**Pull request by antonrom1**
Up until now python tax formula was parsed using simple regex with custom token parsing. This method had its limits tho, such as you could with no issue generate ast nodes such as tuples, for example which would not work with the JS tax. This commit is a more permanent solution with actual AST parsing and validation with a whitelist-only approach.

In this commit we do a two-step pass on formulas with ast. 
- First one is the transfomer which simply collects all accessed fields (through subscripts and through attribute access) and rewrites all attributes to subscritps (because 2nd validaiton step does not allow attributes). 
- The second step valides the formula AST against a whitelist of nodes. If any node is not whitelisted explicitly, we will get a validation error. 

This allows us to be much more granular in what we allow or not.

todo:
- [x] test edge cases that we now handle
- [x] fix lazy translation not working (it only goes up to 1 frame `f_back` above current one for `f_locals` containing `env` or `cr`)
- [x] add an ast transformer & collector to stop using regex to replace attribute access by subscript access (e.g. `product.field` -> `product["field"]`

no-task

### CODE REVIEW

File: addons/account_tax_python/tests/test_taxes_computation.py [hunk 1]
```diff
@@ -83,7 +85,7 @@ def test_formula(self):
             product_values={'volume': 100.0},
         )
         self.assert_python_taxes_computation(
-            "product.volume > 100 and 10 or 5",
+            'product["volume"] > 100 and 10 or 5',
```
Comment (smetl) [inline addons/account_tax_python/tests/test_taxes_computation.py:48]: why?

File: addons/account_tax_python/tests/test_taxes_computation.py [hunk 2]
```diff
@@ -95,7 +97,7 @@ def test_formula(self):
             product_values={'volume': 105.0},
         )
         self.assert_python_taxes_computation(
-            "product.volume > 100 and 10 or 5",
+            "product['volume'] > 100 and 10 or 5",
```
Comment (smetl) [inline addons/account_tax_python/tests/test_taxes_computation.py:60]: why?

File: addons/account_tax_python/tests/test_taxes_computation.py (hunk 1 above)
Comment (antonrom1) [reply addons/account_tax_python/tests/test_taxes_computation.py:48]: attribute access to fields was tested in the previous test, this is just to also cover subscript with double quotes without polluting tests too much

File: addons/account_tax_python/tests/test_taxes_computation.py (hunk 2 above)
Comment (antonrom1) [reply addons/account_tax_python/tests/test_taxes_computation.py:60]: same as above, just wanted to also test simple quotes

**Comment by antonrom1** [comment]
@odoo/rd-security hi 👋

Could you please have a look at this? We are trying to replace the python tax formula syntax validation with AST instead of regex. 
Thank you!

**Comment by xmo-odoo** [comment]
At this point why not build a dedicated parser and evaluator for *exactly* the language of formulas? It's just a handful of operators and a pair of hard-wired functions.

**Comment by smetl** [comment]
@antonrom1 What about targeting 18.0? I have some need in 18.0 to extract the base and amount from the formula. Like "4.6555 * quantity" or "4.6555 * product.turlututu", I need to retrieve 4.6555 & quantity or product.turlututu. I have a branch to do it using regexes but it would be better with this AST to be more accurate

**Comment by antonrom1** [comment]
@xmo-odoo I'm not sure I fully understand your suggestion, could you clarify a bit please?. Wouldn’t a dedicated parser end up being more complex to implement correctly? We would also need it to behave exactly the same both in Python and JS, which might be tricky... I guess it could break some existing Python formulas too

**Comment by xmo-odoo** [comment]
> Wouldn’t a dedicated parser end up being more complex to implement correctly? 

Not really? From the language whitelisted here it's not like the formulas language is very complicated, and that would ensure you have exactly what you want, nor would you need to whitelist then patch up the AST.

> We would also need it to behave exactly the same both in Python and JS, which might be tricky... 

It would be much easier with a custom parser since you'd control the parsed language exactly.

**Review by smetl — APPROVED** [review]

**Review by xmo-odoo — COMMENTED** [review]
@robodoo override=ci/security

**Review by smetl — APPROVED** [review]
@robodoo r+
